clippy_lints/src/lossy_float_literal.rs

```diff
@@ -0,0 +1,85 @@
+use crate::utils::span_lint_and_sugg;
+use crate::utils::sugg::format_numeric_literal;
+use if_chain::if_chain;
+use rustc::ty;
+use rustc_errors::Applicability;
+use rustc_hir as hir;
+use rustc_lint::{LateContext, LateLintPass};
+use rustc_session::{declare_lint_pass, declare_tool_lint};
+use std::{f32, f64};
+use syntax::ast::*;
+
+declare_clippy_lint! {
+    /// **What it does:** Checks for whole number float literals that
+    /// cannot be represented as the underlying type without loss.
+    ///
+    /// **Why is this bad?** Rust will silently lose precision during
+    /// conversion to a float.
+    ///
+    /// **Known problems:** None.
+    ///
+    /// **Example:**
+    ///
+    /// ```rust
+    /// // Bad
+    /// let _: f32 = 16_777_217.0; // 16_777_216.0
+    ///
+    /// // Good
+    /// let _: f32 = 16_777_216.0;
+    /// let _: f64 = 16_777_217.0;
+    /// ```
+    pub LOSSY_FLOAT_LITERAL,
```

Comment:
I'd put this in the same LintPass as the `excessive_precission` lint, to share some code. You can then maybe change the name of the `excessive_precision.rs` module to `float_literals.rs`.

Reply:
Makes sense. Will do.